File: test/js/bun/ffi/bir/fixtures/exports/narrow-parameters-past-the-registers.ts
import { JSCallback } from "bun:ffi";
import * as c from "./narrow-parameters-past-the-registers.c";

console.log(String(c.mixed(1, 2, 3, 4, 5, 6, 7, 8, -1, 100, -300, 70000, 120, 2 ** 40, 20000, 1.5, -128)));
// (What does not fit a parameter's type is converted to it as C converts: 255 to a signed char is -1.)
console.log(String(c.bytes(1, 2, 3, 4, 5, 6, 7, 8, -1, 255, -128, 128, 127, 1, 254, 254, -3)));
console.log(String(c.shorts_and_bools(1, 2, 3, 4, 5, 6, 7, 8, -1, true, 65535, false, -32768, -5)));
console.log(String(c.after_the_doubles(1, 2, 3, 4, 5, 6, 7, 8, 0.5, -4, 1.5, -300)));
const callback = new JSCallback(
  (r1, r2, r3, r4, r5, r6, r7, r8, c1, s1, c2, i1, u1) =>
    BigInt(Number(r1) + Number(r8) + c1 * 3 + s1 * 5 + c2 * 7 + i1 * 11 + u1 * 13),
  {
    args: ["i64", "i64", "i64", "i64", "i64", "i64", "i64", "i64", "i8", "i16", "i8", "i32", "u8"],
    returns: "i64",
  },
);
console.log(String(c.through_a_callback(callback.ptr)));
callback.close();

File: test/js/bun/ffi/bir/fixtures/exports/narrow-parameters-past-the-registers.c
// Functions JavaScript calls whose ninth and later parameters are narrower than an int: no register is left for
// them, and where they go on the stack is the target's business (Apple's arm64 packs them at their own size). What
// arrives is the value JavaScript passed, converted as the parameter's type says, whatever lies next to it.
long long mixed(long long r1, long long r2, long long r3, long long r4, long long r5, long long r6, long long r7, long long r8, char c1, char c2, short s1, int i1, char c3, long long l1, short s2, float f1, char c4) {
  return r1 + r8 + c1 * 3LL + c2 * 5LL + s1 * 7LL + i1 * 11LL + c3 * 13LL + l1 * 17LL + s2 * 19LL + (long long)(f1 * 23) + c4 * 29LL;
}
long long bytes(long long r1, long long r2, long long r3, long long r4, long long r5, long long r6, long long r7, long long r8, signed char a, unsigned char b, signed char c, unsigned char d, signed char e, unsigned char f, signed char g, unsigned char h, signed char i) {
  return r2 + a + b * 2LL + c * 3LL + d * 4LL + e * 5LL + f * 6LL + g * 7LL + h * 8LL + i * 9LL;
}
long long shorts_and_bools(long long r1, long long r2, long long r3, long long r4, long long r5, long long r6, long long r7, long long r8, short a, _Bool b, unsigned short c, _Bool d, short e, char f) {
  return r3 + a + b * 2LL + c * 4LL + d * 8LL + e * 16LL + f * 32LL;
}
long long after_the_doubles(double d1, double d2, double d3, double d4, double d5, double d6, double d7, double d8, float f1, char c1, float f2, short s1) {
  return (long long)(d1 + d8 + f1 * 3 + f2 * 5) + c1 * 7LL + s1 * 11LL;
}
// And C calling back into JavaScript with as many.
long long through_a_callback(long long (*callback)(long long, long long, long long, long long, long long, long long, long long, long long, char, short, char, int, unsigned char)) {
  return callback(1, 2, 3, 4, 5, 6, 7, 8, -1, -300, 100, -70000, 200);
}
